### DISCUSSION TITLE
Combining Transcription with Diarization (speaker identification)

### DISCUSSION THREAD

**Discussion by MustafaCQN** [SYSTRAN/faster-whisper#99] [Q&A]
Hi everyone!

I was working on a project that will takes audio file and transcribes the meeting. 

The problem with this project that I am facing is I need to diarize the speakers with their names. I was using Pyannote package for the identification of speakers but the problem is, Both transcription and diarization uses different models which outcomes different timecodes. Because of the different timecodes, I cannot able to match the transcription with the speaker names.

Anybody knows how I can tweak this problem or is there a product/model/method that I can use for both transcribing the matching with the speaker name using timecodes?

![image](https://user-images.githubusercontent.com/16370078/229106847-5ec883a6-6231-42dc-9ae2-b711ef364c39.png)
Left is speaker identifications with timecodes (pyannote), right is the transcription with timecodes (faster-whisper)

Thank you!

**Comment by apzl** [comment]
Have you checked out [WhisperX](https://github.com/m-bain/whisperX/tree/main)?

**Reply by ciekawy** [reply]
its slower. Can faster-whisper be used within WhisperX?

**Reply by landemou** [reply]
it would be great to have the diarization on faster-whisper but surely very hard to set up !

**Reply by MustafaCQN** [reply]
> Have you checked out [WhisperX](https://github.com/m-bain/whisperX/tree/main)?

I will check and reply here after I see the results. Right now I am using another Thread to check speakers speaking identification from google meet. But as you guess its just solving the problem from one platform. And It can't be used cases like phone recording of real meetings etc.



> it would be great to have the diarization on faster-whisper but surely very hard to set up !

I would love if faster-whisper releases this!. I was using Pyannote then manual scraping from google meet but in the end both of them giving me different timecodes than faster-whisper. So I have to combine them together for understandable diarization. I have solved this after I widen both timecodes to the nexts beginning code.
Ex: 
[5.0 -> **_10.0_**] hi
[**_12.0_** -> 14.0] how are you

to
[5.0 -> **_12.00_**] hi
[_**12.0**_ -> 14.0] how are you

then I take the average of the start and end time of the transcription and directly took the speaker name from the given diarization list.
That way this will create a diarization with approx. %80 accuracy with using 2 different models or 1 model and 1 automation thread.

**Reply by ciekawy** [reply]
Shouldn't be the same compose of pyannote on whisper as in whisperX /  pyannote-whisper project?

**Comment by NavodPeiris** [comment]
checkout this repo: https://github.com/Navodplayer1/speechlib
this uses pyannote diarization and segment the audio according to start and end times. Then apply faster-whisper transcription to each segment. Finally output transcript with **time from pyannote diarization** and **transcripted text from faster-whisper**. 

**you will get accurate timing**

You can also do speaker recognition if you provide voices_folder. Then transcription will contain actual speaker names!

**Reply by RustX2802** [reply]
Hi @NavodPeiris, is it possible to apply **diarization** with **speechlib** for **real-time** transcription capabilities? Have you tried this option?

**Comment by kinthaiofficial** [comment]
Combining transcription with diarization is a great foundation for voice-aware multi-agent systems — once you know who said what, you can route different speakers' utterances to different agents or use speaker identity as an access control signal.

A few practical patterns for the transcription + diarization pipeline:

**Pyannote + faster-whisper** — the most reliable open-source stack. Pyannote handles speaker timestamps (who spoke when), faster-whisper transcribes the segments. The key is using the pyannote timestamps to slice the audio for whisper transcription rather than letting whisper segment independently.

**Word-level alignment** — after transcription, use whisper's `word_timestamps=True` and align the word-level timestamps with pyannote's speaker timestamps. This gives you speaker attribution per word, not just per segment.

**Speaker embedding for identification** — beyond diarization (distinguishing speakers A from B), speaker verification (is this the registered user "Alice"?) requires embedding comparison. ECAPA-TDNN embeddings work well for this and can run on CPU.

For agent systems, voice-based speaker identity creates an interesting auth pattern: a registered voice as a "soft" auth factor. Not strong enough as sole auth, but useful for "this request sounds like it's from the account owner, not an intruder."

We've been thinking about voice-authenticated agent commands as a modality in KinthAI — the identity layer that makes this possible: https://blog.kinthai.ai/221-agents-multi-agent-coordination-lessons

Are you using this for meeting transcription, conversation agents, or accessibility tooling?

**Comment by paramashivatma** [comment]
Unsubscribe

**Comment by etlweather** [comment]
I did implement something with speaker ID using Pyannote + faster-whisper and use the Word timestamp. The only difficulty I am running into is faster-whisper is basically second-based timestamp so when people talk quickly in turn, some utterances are left with the wrong speaker. Used with meetings.

**Reply by offbyseven** [reply]
I’m building a meeting note taking app too and ran into similar issues with pyannote and whisper, it seems like no matter what I did some utterances would be labeled incorrectly.

I think the issue comes down to machine diarization having to account for things like people taklking over each other and pauses. I’m looking into [perfect diarization](https://www.recall.ai/blog/speaker-diarization#1-perfect-diarization) atm which uses separate audio streams to label who said what. Looks promising and is what my team is trying next

**Comment by praveenperera** [comment]
For diarization you can try https://github.com/avencera/speakrs, would be a good fit especially for users on macs
